### PR TITLE
Ocs independent mode external cluster init

### PR DESCRIPTION
**Pull request by aruniiird**
Support for rook external cluster mode in OCS Operator

Modifies the existing StorageCluster's `Reconcile` function to support
ExternalCluster mode.

### CODE REVIEW

File: pkg/apis/ocs/v1/storagecluster_types.go [hunk 1]
```diff
@@ -125,16 +125,24 @@ type NodeTopologyMap struct {
...
 const (
-	ReconcileFailed           = "ReconcileFailed"
-	ReconcileInit             = "Init"
-	ReconcileCompleted        = "ReconcileCompleted"
-	ReconcileCompletedMessage = "Reconcile completed successfully"
+	ReconcileFailed                 = "ReconcileFailed"
+	ReconcileInit                   = "Init"
+	ReconcileCompleted              = "ReconcileCompleted"
+	ReconcileCompletedMessage       = "Reconcile completed successfully"
+	ExternalClusterConnected        = "ExternalClusterConnected"
```
Comment (umangachapagain) [inline pkg/apis/ocs/v1/storagecluster_types.go]: This is a Phase. There is a separate place for this.

Comment (umangachapagain) [reply pkg/apis/ocs/v1/storagecluster_types.go]:
```suggestion
	ExternalClusterConnected        = "Connected"
```

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 2]
```diff
@@ -63,6 +108,119 @@ func (r *ReconcileStorageCluster) ReconcileExternalStorageCluster(sc *ocsv1.Stor
...
+		// to set readiness.
+		r := ready.NewFileReady()
+		err = r.Set()
+		if err != nil {
+			reqLogger.Error(err, "Failed to mark operator ready")
+			return reconcile.Result{}, err
+		}
+		if sc.Status.Phase != statusutil.PhaseClusterExpanding {
+			sc.Status.Phase = statusutil.PhaseReady
+		}
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]:
```suggestion

```

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 3]
```diff
@@ -63,6 +108,119 @@ func (r *ReconcileStorageCluster) ReconcileExternalStorageCluster(sc *ocsv1.Stor
...
+		// If for any reason we marked ourselves !upgradeable...then unset readiness
+		if conditionsv1.IsStatusConditionFalse(sc.Status.Conditions, conditionsv1.ConditionUpgradeable) {
+			r := ready.NewFileReady()
+			err = r.Unset()
+			if err != nil {
+				reqLogger.Error(err, "Failed to mark operator unready")
+				return reconcile.Result{}, err
+			}
+		}
+		if sc.Status.Phase != statusutil.PhaseClusterExpanding {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]: We don't have Phase Expanding here.

File: pkg/apis/ocs/v1/storagecluster_types.go (hunk 1 above)
Comment (aruniiird) [reply pkg/apis/ocs/v1/storagecluster_types.go]: @umangachapagain , What I added was a condition, not a phase... Do you still think, we need phase here not condition?

**Comment by aruniiird** [comment]
@jarrpa , @leseb , @travisn , please take look...

File: deploy/olm-catalog/ocs-operator/4.3.0/ocs-operator.v4.3.0.clusterserviceversion.yaml [hunk 4]
```diff
@@ -1155,7 +1156,7 @@ spec:
                   value: noobaa/noobaa-core:5.2.11
                 - name: NOOBAA_DB_IMAGE
                   value: centos/mongodb-36-centos7
-                image: quay.io/ocs-dev/ocs-operator:latest
+                image: quay.io/aruniiird/ocs-operator:amohans-ocs-operator-independent-mode-init-no-phase-v1
```
Comment (umangachapagain) [inline deploy/olm-catalog/ocs-operator/4.3.0/ocs-operator.v4.3.0.clusterserviceversion.yaml]:
```suggestion
                image: quay.io/ocs-dev/ocs-operator:latest
```

File: pkg/controller/defaults/resources.go [hunk 5]
```diff
@@ -11,72 +11,72 @@ var (
...
-				corev1.ResourceCPU:    resource.MustParse("2"),
-				corev1.ResourceMemory: resource.MustParse("4Gi"),
+				corev1.ResourceCPU:    resource.MustParse("0"),
+				corev1.ResourceMemory: resource.MustParse("0Gi"),
 			},
 			Limits: corev1.ResourceList{
-				corev1.ResourceCPU:    resource.MustParse("2"),
-				corev1.ResourceMemory: resource.MustParse("4Gi"),
+				corev1.ResourceCPU:    resource.MustParse("0"),
+				corev1.ResourceMemory: resource.MustParse("0Gi"),
```
Comment (umangachapagain) [inline pkg/controller/defaults/resources.go]: Do not commit these changes.

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+	// in-memory conditions should start off empty. It will only ever hold
+	// negative conditions (!Available, Degraded, Progressing)
+	r.conditions = nil
+	// Start with empty r.phase
+	r.phase = ""
+
+	for _, f := range []func(*ocsv1.StorageCluster, logr.Logger) error{
+		// Add support for additional resources here
+		// r.ensureStorageClasses,
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]: If this is not required, remove it.

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+		r.ensureExternalCephCluster,
+		r.ensureNoobaaSystem,
+	} {
+		err = f(sc, reqLogger)
+		if r.phase == statusutil.PhaseClusterExpanding {
+			sc.Status.Phase = statusutil.PhaseClusterExpanding
+			phaseErr := r.client.Status().Update(context.TODO(), sc)
+			if phaseErr != nil {
+				reqLogger.Error(phaseErr, "Failed to set PhaseClusterExpanding")
+			}
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]: External Storage Cluster does not expand.

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+			Namespace: sc.Namespace,
+			Labels:    labels,
+		},
+		Spec: cephv1.ClusterSpec{
+			External: cephv1.ExternalSpec{
+				Enable: true,
+			},
+			DataDirHostPath: "/var/lib/rook",
+			CephVersion: cephv1.CephVersionSpec{
+				Image: cephImage,
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]:
```suggestion
```
We don't need CephVersion for External Ceph Cluster.

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+		// here negative conditions for external cluster has tobe set
+	}
+
+	// When phase is expanding, wait for CephCluster state to be updating
+	// this means expansion is in progress and overall system is progressing
+	// else expansion is not yet triggered
+	if sc.Status.Phase == statusutil.PhaseClusterExpanding &&
+		found.Status.State != cephv1.ClusterStateUpdating {
+		r.phase = statusutil.PhaseClusterExpanding
+	}
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]:
```suggestion

```

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/reference"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ReconcileExternalStorageCluster reconciles the external Storage Cluster
+func (r *ReconcileStorageCluster) ReconcileExternalStorageCluster(sc *ocsv1.StorageCluster) (reconcile.Result, error) {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]: Need a check to see if another StorageCluster Object is already present.
If converged StorageCluster is present, it can not be replaced with External StorageCluster and vice-versa.

File: deploy/olm-catalog/ocs-operator/4.3.0/ocs-operator.v4.3.0.clusterserviceversion.yaml (hunk 4 above)
Comment (aruniiird) [reply deploy/olm-catalog/ocs-operator/4.3.0/ocs-operator.v4.3.0.clusterserviceversion.yaml]: Changed it to latest operator image..

File: pkg/controller/defaults/resources.go (hunk 5 above)
Comment (aruniiird) [reply pkg/controller/defaults/resources.go]: Thanks Umanga, reverted the changes...

File: pkg/controller/storagecluster/externalstoragecluster_controller.go (hunk 6 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/externalstoragecluster_controller.go]: Removed...

File: pkg/controller/storagecluster/externalstoragecluster_controller.go (hunk 7 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/externalstoragecluster_controller.go]: Removed the lines...

File: pkg/controller/storagecluster/externalstoragecluster_controller.go (hunk 8 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/externalstoragecluster_controller.go]: Hope we don't need the `cephImage` as in the comment suggests,
https://github.com/rook/rook/blob/ce1884388543f4dfa445ae6a66a77cb65441d678/cluster/examples/kubernetes/ceph/cluster-external.yaml#L22

File: pkg/controller/storagecluster/externalstoragecluster_controller.go (hunk 9 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/externalstoragecluster_controller.go]: Removed...

File: pkg/controller/storagecluster/externalstoragecluster_controller.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+	// negative conditions (!Available, Degraded, Progressing)
+	r.conditions = nil
+	// Start with empty r.phase
+	r.phase = ""
+
+	for _, f := range []func(*ocsv1.StorageCluster, logr.Logger) error{
+		// Add support for additional resources here
+		r.ensureCephConfig,
+		r.ensureExternalCephCluster,
+		r.ensureNoobaaSystem,
```
Comment (anmolsachan) [inline pkg/controller/storagecluster/externalstoragecluster_controller.go]: it would be better if we comment it out here in this PR and send a separate PR for noobaa work.

**Comment by obnoxxx** [comment]
@aruniiird, could you please explain more what the PR does in the PR description?

The title

> Ocs independent mode external cluster init

and especially the description text

> This is a PR on top of Umanga's (PR), and consists mainly the same code as the main reconcile method.

don't really describe at all what this trying to achieve. Please describe briefly what "Umanga's PR" (please reference it) has done and what is the gap that this PR closes. Possibly (if helpful) describe how it's done.

File: pkg/controller/storagecluster/externalstoragecluster_controller.go (hunk 10 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/externalstoragecluster_controller.go]: Now we are making changes into the original `Reconcile` method, the said functionality is already present

File: pkg/controller/storagecluster/externalstoragecluster_controller.go (hunk 11 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/externalstoragecluster_controller.go]: Done...

**Comment by aruniiird** [comment]
> @aruniiird, could you please explain more what the PR does in the PR description?
> 
> The title
> 
> > Ocs independent mode external cluster init
> 
> and especially the description text
> 
> > This is a PR on top of Umanga's (PR), and consists mainly the same code as the main reconcile method.
> 
> don't really describe at all what this trying to achieve. Please describe briefly what "Umanga's PR" (please reference it) has done and what is the gap that this PR closes. Possibly (if helpful) describe how it's done.

@obnoxxx , Thank you for your pointers... I have changed the commit message (as well as the PR's top comment), will that be good enough.

**Comment by aruniiird** [comment]
Closing this PR as @vbnrh , has all the changes (including this) added to his PR...
https://github.com/openshift/ocs-operator/pull/429
